connectors/sources/jira.py

```diff
@@ -0,0 +1,467 @@
+
+        Yields:
+            Dictionary: Jira attachment on the given issue
+            attachment (dict): Attachment dictionary for extracting the content
+        """
+        for attachment in attachments:
+            yield {
+                "_id": f"{issue_key}-{attachment['id']}",
+                "title": attachment["filename"],
+                "type": "Attachment",
+                "issue": issue_key,
+                "_timestamp": attachment["created"],
+                "size": attachment["size"],
+            }, attachment
+
+    async def _grab_content(self, attachments, issue_key):
+        """Coroutine to add attachments to Queue and get content
+
+        Args:
+            attachments (list): List of attachments for an issue
+            issue_key (str): Issue key for generating `_id` field
+        """
+        async for content, attachment in self._get_attachments(
+            attachments=attachments, issue_key=issue_key
+        ):
+            if attachment:
+                await self.queue.put(
+                    (
+                        content,
+                        partial(
+                            self.get_content,
+                            issue_key=issue_key,
+                            attachment=copy(attachment),
+                        ),
+                    )
+                )
+
+    async def get_docs(self, filtering=None):
+        """Executes the logic to fetch jira objects in async manner
+
+        Args:
+            filtering (Filtering): Object of class Filtering
+
+        Yields:
+            dictionary: dictionary containing meta-data of the files.
+        """
+
+        async def _project_task():
+            """Coroutine to add projects documents to Queue"""
+            async for project_data in self._get_projects():
+                await self.queue.put((project_data, None))
+            await self.queue.put("FINISHED")
+
+        async def _document_task():
+            """Coroutine to add issues/attachments to Queue"""
+            async for document, issue in self._get_issues():
+                await self.queue.put((document, None))
+                attachments = issue["fields"]["attachment"]
+                if len(attachments) > 0:
+                    await self.fetchers.put(
+                        partial(self._grab_content, attachments, issue["key"])
+                    )
+            await self.queue.put("FINISHED")
+
+        project_task = asyncio.create_task(_project_task())
+        attachment_task = asyncio.create_task(_document_task())
+        completed_tasks = 0
+        while True:
+            _, item = await self.queue.get()
+            if item == "FINISHED":
+                completed_tasks += 1
+                if completed_tasks == 2:
+                    break
+            else:
+                yield item
+
+        await self.fetchers.join()
+        await asyncio.gather(project_task, attachment_task)
```

Comment:
Is this line needed? If you exit the infinite loop above, does it mean both `project_task` and `attachment_task` are done?

Reply:
Yes, the consumer stops when it gets `FINISHED` twice from the queue and that is going to happen at the time when both tasks are done.